sklearn/metrics/_classification.py

```diff
@@ -555,77 +564,106 @@ def multilabel_confusion_matrix(
-        pred_sum = pred_sum[indices]
+        indices = _searchsorted(xp, sorted_labels, labels[:n_labels])
+        tp_sum = xp.take(tp_sum, indices, axis=0)
+        true_sum = xp.take(true_sum, indices, axis=0)
+        pred_sum = xp.take(pred_sum, indices, axis=0)
 
     else:
         sum_axis = 1 if samplewise else 0
 
         # All labels are index integers for multilabel.
         # Select labels:
-        if not np.array_equal(labels, present_labels):
-            if np.max(labels) > np.max(present_labels):
+        if labels.shape != present_labels.shape or xp.any(
+            xp.not_equal(labels, present_labels)
+        ):
+            if xp.max(labels) > xp.max(present_labels):
                 raise ValueError(
                     "All labels must be in [0, n labels) for "
                     "multilabel targets. "
-                    "Got %d > %d" % (np.max(labels), np.max(present_labels))
+                    "Got %d > %d" % (xp.max(labels), xp.max(present_labels))
                 )
-            if np.min(labels) < 0:
+            if xp.min(labels) < 0:
                 raise ValueError(
                     "All labels must be in [0, n labels) for "
                     "multilabel targets. "
-                    "Got %d < 0" % np.min(labels)
+                    "Got %d < 0" % xp.min(labels)
                 )
 
         if n_labels is not None:
             y_true = y_true[:, labels[:n_labels]]
             y_pred = y_pred[:, labels[:n_labels]]
 
         # calculate weighted counts
-        true_and_pred = y_true.multiply(y_pred)
-        tp_sum = count_nonzero(
-            true_and_pred, axis=sum_axis, sample_weight=sample_weight
-        )
-        pred_sum = count_nonzero(y_pred, axis=sum_axis, sample_weight=sample_weight)
-        true_sum = count_nonzero(y_true, axis=sum_axis, sample_weight=sample_weight)
+        if _is_numpy_namespace(xp=xp):
+            true_and_pred = y_true.multiply(y_pred)
+            tp_sum = count_nonzero(
+                true_and_pred, axis=sum_axis, sample_weight=sample_weight
+            )
+            pred_sum = count_nonzero(y_pred, axis=sum_axis, sample_weight=sample_weight)
+            true_sum = count_nonzero(y_true, axis=sum_axis, sample_weight=sample_weight)
+        else:
+            true_and_pred = xp.multiply(y_true, y_pred)
+            tp_sum = _count_nonzero(
+                true_and_pred,
+                xp=xp,
+                device=device_,
+                axis=sum_axis,
+                sample_weight=sample_weight,
+            )
+            pred_sum = _count_nonzero(
+                y_pred,
+                xp=xp,
+                device=device_,
+                axis=sum_axis,
+                sample_weight=sample_weight,
+            )
+            true_sum = _count_nonzero(
+                y_true,
+                xp=xp,
+                device=device_,
+                axis=sum_axis,
+                sample_weight=sample_weight,
+            )
```

Comment:
Is the only difference between the two branches the `device=device_` keyword argument passing?

Why not just drop the NumPy-specific branch? The array API branch should always work, no?

Reply:
The first branch is specifically for handling sparse arrays. That was the default before introducing the array api and thats why we have this split.

Reply:
The count_nonzero is the function that was already present by default and it handles sparse arrays. The _count_nonzero is our own method because the array api code is handled differently from the sparse cases.

Reply:
but if our `_count_nonzero` handles sparse (which we can make it to do), then we don't need two code branches here.

Reply:
Added the conditioning in the _count_nonzero.